Rewrite the signed-commit composite action to use native git commands

mdaleo-fpf opened the pull request:
Works towards #26 

I removed the _"Adapted from"_ statement at the top as this version (and to be fair even the previous one) is very different now.

This one was a bit of a challenge but _hopefully_ I matched the requirements.

## Test Plan

- [ ] TBD (ensure necessary changes for all the downstream workflows that use this action are staged and themselves ready to merge)

mdaleo-fpf committed:
Rewrite the signed-commit composite action to use native git commands

mdaleo-fpf committed:
Fix white spaces and line lenght to make linter happy

mig5 commented:
Yes, I was keeping this open deliberately til @adaFPF and @enpaul were back from conference. 

Mostly for:

1) the security context around whether we consider `write to GITHUB_ENV may allow code execution` a realistic problem for us here (being a public repo I guess) - one for @adaFPF 

2) and because @enpaul (who I consider more of an Actions expert than I am) might have an idea on how it could be avoided or if it's needed (is it because $COMMIT_SHA isn't otherwise 'retained' across steps in the same action workflow?)

enpaul commented on act/signed-commit/action.yml:
```diff
@@ -29,30 +23,77 @@ inputs:
+
+        echo "$PGP_KEY_B64" | base64 --decode | gpg --import
+
+        KEY_ID=$(
```
Is this env var meant to be output for usage in later steps? I think this is just living in the current job env at the moment

enpaul commented on act/signed-commit/action.yml:
```diff
@@ -29,30 +23,77 @@ inputs:
-    required: true
+outputs:
+  commit-sha:
+    description: SHA of the created commit
```
To make these outputs functional I think you'll need to specify a `value:` field here with a reference to the step output that should populate them. This is to "move" the output from the step scope to the action scope. An example of this is [here](https://github.com/freedomofpress/actionslib/blob/main/act/delete-artifact/action.yml#L30)

enpaul commented on act/signed-commit/action.yml:
```diff
@@ -29,30 +23,77 @@ inputs:
+        git config user.signingkey "$GPG_KEY_ID"
+        git config commit.gpgsign true
+        git config tag.gpgsign true
+
```
Also someone feel free to chime in if you think this is bad practice. This feels a bit like verging on making a monolithic action that handles all things git, which is not at all what I'd want ([ref](https://en.wikipedia.org/wiki/Unix_philosophy#Origin)). Welcome to hear dissenting opinions here.

enpaul reviewed (changes requested):
Thanks so much for putting this together @mdaleo-fpf ! Is there an easy-ish way we could do a test run of the action just to confirm that the commit turns out signed the way we expect?

mig5 commented on act/signed-commit/action.yml:
```diff
@@ -29,30 +23,77 @@ inputs:
+        git commit -m "$MSG"
+
+        SHA=$(git rev-parse HEAD)
+        echo "COMMIT_SHA=$SHA" >> $GITHUB_ENV
```
That sounds good! And therefore @mdaleo-fpf I think we'd want to do the same technique over in the weblate PR per my comms in https://github.com/freedomofpress/infrastructure/pull/6145#discussion_r2540760852

mdaleo-fpf committed:
Edit outputs and add feature branch creation, remove redundant final step to emit outputs

mdaleo-fpf committed:
Add composite action to determine latest git tag

mdaleo-fpf committed:
Eliminate template injection warnings by moving inputs into env vars

enpaul committed:
Fix yamllint line length warning

enpaul commented:
> @enpaul I pushed some changes based on your comments. yamllint is very strict here with the line length

Looks good @mdaleo-fpf!

For the yamllint complaints, IMO this is a good place to just disable that warning. In general, I'm a fan of disabling linters where multiple languages (or, in this case, semantic markup structures) collide. Zizmor is already independently linting the actions fields and templates, and breaking up the line to satisfy the length limit would be a net loss for readability and potentially even introduce bugs because Github is not great at parsing their own YAML spec.

Yamllint is overkill for most repositories and (like here) causes more problems than it's worth. I added it to this one just because there's so much yaml in this repo, but I'm open to ditching it too. I'd welcome any insight on that.

I've pushed a commit with the line length warning disabled for the offending line.

enpaul committed:
Add logic for automatically adding 'latest' tag when on the default branch

mdaleo-fpf committed:
Add test-publish-r2 workflow

enpaul committed:
Update oci-build workflow to automatically determine default branch

By determining the default branch from the github API we save one more user input
that requires updating

enpaul committed:
Replace custom image metadata labels with standard opencontainer labels

enpaul committed:
Replace dedicated push job with conditional push step

mig5 committed:
Update publish-r2 flow to take an optional input for artifact downloading (used by integration test)

mig5 committed:
Add explicit permissions to resolve zizmor issues

mig5 committed:
Run on changes to the test-publish-r2.yml as well

mig5 committed:
rename file

apyrgio committed:
Publish artifacts to a directory in an R2 bucket

Extend the `publish-r2.yaml` action with an extra input argument
(`target_dir`), that specifies the directory in the target bucket that
it should publish artifacts to. By default, this argument is empty,
which means that the action will publish the artifacts to the root
directory of the bucket.

enpaul committed:
Add required inputs and secrets to signed-commit action

Composite actions don't support accessing variables or secrets from the caller's
namespace and so need all inputs and secrets passed into them
